service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/mapper/RequestStatementMapper.java

```diff
@@ -41,38 +41,101 @@
-            requestStatementTemplateParameters.setDefaultReasonCode(DEFAULT_REASON_CODE_XML);
+            requestStatementTemplateParameters.defaultReasonCode(DEFAULT_REASON_CODE_XML);
         }
 
-        return TemplateUtils.fillTemplate(REQUEST_STATEMENT_TEMPLATE, requestStatementTemplateParameters);
+        if (referralRequest.hasRequester()
+                && referralRequest.getRequester().hasAgent()
+                && referralRequest.getRequester().getAgent().hasReference()) {
+            var requester = referralRequest.getRequester();
+            Reference agentRef = requester.getAgent();
+            var onBehalfOf = requester.hasOnBehalfOf() ? requester.getOnBehalfOf() : null;
+            processAgent(agentRef, onBehalfOf, requestStatementTemplateParameters);
+        }
+
+        if (referralRequest.hasRecipient()) {
+            final Reference recipient = referralRequest.getRecipientFirstRep();
+            final var responsibleParty = idMapper.get(recipient);
+            requestStatementTemplateParameters.responsibleParty(responsibleParty);
+        }
+
+        return TemplateUtils.fillTemplate(REQUEST_STATEMENT_TEMPLATE, requestStatementTemplateParameters.build());
+    }
+
+    private void processAgent(Reference agent, Reference onBehalfOf,
+                              RequestStatementTemplateParametersBuilder requestStatementTemplateParameters) {
+        final IdMapper idMapper = messageContext.getIdMapper();
+
+        if (isReferenceToType(agent, ResourceType.Practitioner) && isReferenceToType(onBehalfOf, ResourceType.Organization)) {
+            final String participantRef = idMapper.get(onBehalfOf);
+            final String participant = participantMapper.mapToParticipant(participantRef, ParticipantType.AUTHOR);
+            requestStatementTemplateParameters.participant(participant);
+
+        } else if (isReferenceToType(agent, ResourceType.Practitioner) || isReferenceToType(agent, ResourceType.Organization)) {
+            final String participantRef = idMapper.getOrNew(agent);
+            final String participant = participantMapper.mapToParticipant(participantRef, ParticipantType.AUTHOR);
+            requestStatementTemplateParameters.participant(participant);
+
+        } else if (isReferenceToType(agent, ResourceType.Device)) {
+            messageContext.getInputBundleHolder().getResource(agent.getReferenceElement())
+                .filter(Device.class::isInstance)
+                .map(Device.class::cast)
+                .filter(Device::hasType)
+                .map(Device::getType)
+                .flatMap(CodeableConceptMappingUtils::extractTextOrCoding)
+                .map("Requester Device: "::concat)
+                .ifPresentOrElse(requestStatementTemplateParameters::text, () -> {
+                    throw new EhrMapperException("Could not resolve Device Reference");
+                });
+
+        } else if (isReferenceToType(agent, ResourceType.Organization)) {
```

Comment:
Would we ever reach this? Line 95 has the same check

Reply:
Well spotted! I think actually those first two conditions are separate from the rest and the chain could be broken there. Then just need to make sure the exception is thrown in the right conditions